app/services/identity_linker.rb

```diff
@@ -1,17 +1,28 @@
-IdentityLinker = Struct.new(:user, :provider) do
-  attr_reader :identity
+class IdentityLinker
+  attr_reader :user, :provider, :identity
+
+  def initialize(user, provider)
+    @user = user
+    @provider = provider
+  end
 
   def link_identity(session_uuid: nil, **extra_attrs)
     attributes = merged_attributes(session_uuid, extra_attrs)
     identity.update!(attributes)
   end
 
+  def already_linked?
+    identity_relation.exists?
+  end
+
   private
 
   def identity
-    @identity ||= Identity.find_or_create_by(
-      service_provider: provider, user_id: user.id
-    )
+    @identity ||= identity_relation.first_or_create
+  end
+
+  def identity_relation
+    user.identities.where(service_provider: provider)
```

Comment:
It's still memoized in the `identity` method, I just wanted a way to share the query for between `identity` and `already_linked?`, so I extracted a shared helper method.

I don't think the `already_exists?` method needs to be memoized.